### PR TITLE
Build artifact namespacing

### PR DESCRIPTION
**Pull request by kevingranade**
Pretty self-documenting.
Addresses #17339 
tl;dr the payoff is you can have the Makefile stash your object files in different directories, and give your executable files different names so when you switch from tiles to curses or debug to release, you don't have to delete all the intermediate files (and rebuilds will stay faster).

As a side effect, it also means you can run multiple builds simultaneously against the same directory if you are so inclined and think builds are too fast :p

I hacked this together in the middle of the night, and my make-fu isn't nearly as good as my c++-fu, so please review if you know make and test if you don't, and either way report some results here.
I feel fairly good about it not breaking things if you don't enable it at least.

Also if someone has some cleaner ways to do things, I'm all ears.

### CODE REVIEW

**Comment by kevingranade** [comment]
Side note, ccache accomplishes mostly the same thing but in a more flexible and provably correct way, so consider doing that instead.

File: Makefile [hunk 1]
```diff
@@ -854,7 +866,7 @@ else
 	@echo Cannot run an astyle check, your system either does not have astyle, or it is too old.
 endif
 
-tests: version cataclysm.a
+tests: version $(BUILD_PREFIX)cataclysm.a
 	$(MAKE) -C tests
 
 check: version cataclysm.a
```
Comment (BevapDin) [inline Makefile]: Why is this dependency different from the above for "tests"?

**Comment by Coolthulhu** [comment]
Doesn't work well with tests:
```
make BUILD_PREFIX="debug/"

make -C tests
make[1]: Entering directory `/home/lolwut/dda/Cataclysm-DDA/tests'
mkdir debug/obj
mkdir: cannot create directory ‘debug/obj’: No such file or directory
make[1]: *** [debug/obj] Error 1
```

And then for every test:
```
error: unable to open output file 'debug/obj/creature_test.o': 'No such file or directory'
```
```

**Comment by Coolthulhu** [comment]
After manually creating the directory in question, it works fine, though.

File: Makefile (hunk 1 above)
Comment (kevingranade) [inline Makefile]: because I missed it, crap :P

**Comment by kevingranade** [comment]
I'm about 99% sure the problem was mkdir in the tests Makefile not having -p, so it wouldn't create if your prefix is a directory.  I was prefixing with a plain string, so it just changed the name of the directory.
I checked, everywhere else is already using -p.
I also addressed the issue BevapDin pointed out.
